### DISCUSSION TITLE
How discord-player is seeking

### DISCUSSION THREAD

**Discussion by Fyphen1223** [Androz2091/discord-player#1694] [Q&A] [answered]
I want to install seek function on my discordjs only music bot. But I don't know how to do that :(
I looked the code to find the answer, but I can't find it.
Could someone help me?

**Comment by twlite** [comment, marked as answer]
Our current implementation uses ffmpeg to perform seek and to apply some audio filters to the player.

When `playerNode.seek(duration)` is invoked, it internally calls `<GuildQueueAudioFilters>.triggerReplay(duration)`. The implementation for `triggerReplay` is https://github.com/Androz2091/discord-player/blob/ee275af0758372e3f2a3b45b2d96bcbf1eb2f166/packages/discord-player/src/Structures/GuildQueueAudioFilters.ts#L251-L265

Upon further inspection, you can see that `triggerReplay` is just calling `playerNode.play()` with current track as input where queue is set to false and transition mode is set to true for skipping events dispatcher. If you inspect the `playerNode.play()` method, it is calling `this.#createFFmpegStream` to generate a s16le pcm stream from the given source with seek duration https://github.com/Androz2091/discord-player/blob/ee275af0758372e3f2a3b45b2d96bcbf1eb2f166/packages/discord-player/src/Structures/GuildQueuePlayerNode.ts#L462

The implementation of `#createFFmpegStream` is https://github.com/Androz2091/discord-player/blob/ee275af0758372e3f2a3b45b2d96bcbf1eb2f166/packages/discord-player/src/Structures/GuildQueuePlayerNode.ts#L543-L562

This private method is calling `FFmpegFilteter.createStream` which invokes `createFFmpegStream` function which is implemented in this way:
https://github.com/Androz2091/discord-player/blob/ee275af0758372e3f2a3b45b2d96bcbf1eb2f166/packages/discord-player/src/utils/FFmpegStream.ts#L1-L68

Overall, the current implementation of discord-player re-executes ffmpeg with seek duration applied to perform seek.

**Comment by Fyphen1223** [comment]
Thank you!
